wgengine/packet/packet.go

```diff
@@ -6,76 +6,64 @@ package packet
 
 import (
 	"encoding/binary"
+	"errors"
 	"fmt"
-	"log"
-	"net"
 	"strings"
 
 	"tailscale.com/types/strbuilder"
 )
 
-type IPProto int
-
-const (
-	Junk IPProto = iota
-	Fragment
-	ICMP
-	UDP
-	TCP
-)
-
 // RFC1858: prevent overlapping fragment attacks.
 const minFrag = 60 + 20 // max IPv4 header + basic TCP header
 
-func (p IPProto) String() string {
-	switch p {
-	case Fragment:
-		return "Frag"
-	case ICMP:
-		return "ICMP"
-	case UDP:
-		return "UDP"
-	case TCP:
-		return "TCP"
-	default:
-		return "Junk"
-	}
-}
-
-// IP is an IPv4 address.
-type IP uint32
-
-// NewIP converts a standard library IP address into an IP.
-// It panics if b is not an IPv4 address.
-func NewIP(b net.IP) IP {
-	b4 := b.To4()
-	if b4 == nil {
-		panic(fmt.Sprintf("To4(%v) failed", b))
-	}
-	return IP(binary.BigEndian.Uint32(b4))
-}
-
-func (ip IP) String() string {
-	return fmt.Sprintf("%d.%d.%d.%d", byte(ip>>24), byte(ip>>16), byte(ip>>8), byte(ip))
-}
-
-// ICMP types.
-const (
-	ICMPEchoReply    = 0x00
-	ICMPEchoRequest  = 0x08
-	ICMPUnreachable  = 0x03
-	ICMPTimeExceeded = 0x0b
-)
+// Declared here for the lack of a TCPHeader struct.
+const tcpHeaderLength = 20
 
 const (
 	TCPSyn    = 0x02
 	TCPAck    = 0x10
 	TCPSynAck = TCPSyn | TCPAck
 )
 
+var (
+	get16 = binary.BigEndian.Uint16
+	get32 = binary.BigEndian.Uint32
+
+	put16 = binary.BigEndian.PutUint16
+	put32 = binary.BigEndian.PutUint32
+)
+
+var errSmallBuffer = errors.New("buffer too small")
+
+// Header is a packet header capable of marshaling itself into a byte buffer.
+type Header interface {
+	// Length returns the length of the header after marshaling.
+	Length() int
+	// Marshal serializes the header into buf in wire format.
+	// It clobbers the header region, which is the first h.Length() bytes of buf.
+	// It explicitly initializes every byte of the header region,
+	// so pre-zeroing it on reuse is not required. It does not allocate memory.
+	// It fails if and only if len(buf) < Length().
+	Marshal(buf []byte) error
+	// NewPacketWithPayload generates a new packet with the given payload.
```

Comment:
Is this the same code for all implementations? If it can be implemented using only the other methods (Len and Marshal), then perhaps this should just be a func in the package that takes a `Header` interface value?

Reply:
I did that initially (`Generate(header Header, payload []byte)`, but that would lead to `payload` escaping unnecessarily. Though I guess if we're allocating anyway, that may not be a big deal.

Reply:
Hm, yes, that's definitely right. I was probably doing something else stupidly at the time I had it that way. I'll change that.